src/envs/satellites.py

```diff
@@ -0,0 +1,598 @@
+        assert self.action_space.contains(action.arr)
+        # Note: target_sat_x and target_sat_y are only used if we're not
+        # doing calibration, shooting Chemical X or Y, or using an instrument.
+        cur_sat_x, cur_sat_y, obj_x, obj_y, target_sat_x, target_sat_y, \
+            calibrate, shoot_chem_x, shoot_chem_y, use_instrument = action.arr
+        next_state = state.copy()
+        sat = self._xy_to_entity(state, cur_sat_x, cur_sat_y)
+        obj = self._xy_to_entity(state, obj_x, obj_y)
+        if sat is None or obj is None:
+            # Invalid first 4 dimensions of action.
+            return next_state
+        if calibrate > 0.5:
+            # Handle calibration.
+            if not self._Sees_holds(state, [sat, obj]):
+                # Cannot calibrate if the satellite cannot see the object.
+                return next_state
+            if not self._CalibrationTarget_holds(state, [sat, obj]):
+                # Cannot calibrate against the wrong object.
+                return next_state
+            next_state.set(sat, "is_calibrated", 1.0)
+        elif shoot_chem_x > 0.5:
+            # Handle shooting Chemical X.
+            if not self._Sees_holds(state, [sat, obj]):
+                # Cannot shoot if the satellite cannot see the object.
+                return next_state
+            if not self._ShootsChemX_holds(state, [sat]):
+                # Cannot shoot if the satellite doesn't have this chemical.
+                return next_state
+            next_state.set(obj, "has_chem_x", 1.0)
+        elif shoot_chem_y > 0.5:
+            # Handle shooting Chemical Y.
+            if not self._Sees_holds(state, [sat, obj]):
+                # Cannot shoot if the satellite cannot see the object.
+                return next_state
+            if not self._ShootsChemY_holds(state, [sat]):
+                # Cannot shoot if the satellite doesn't have this chemical.
+                return next_state
+            next_state.set(obj, "has_chem_y", 1.0)
+        elif use_instrument > 0.5:
+            # Handle using the instrument on this satellite.
+            if not self._Sees_holds(state, [sat, obj]):
+                # Cannot take a reading if the satellite cannot see the object.
+                return next_state
+            if not self._IsCalibrated_holds(state, [sat]):
+                # Cannot take a reading if the satellite is not calibrated.
+                return next_state
+            if self._HasCamera_holds(state, [sat]) and \
+               not self._HasChemX_holds(state, [obj]):
+                # Cannot take a camera reading without Chemical X.
+                return next_state
+            if self._HasInfrared_holds(state, [sat]) and \
+               not self._HasChemY_holds(state, [obj]):
+                # Cannot take an infrared reading without Chemical Y.
+                return next_state
+            next_state.set(sat, "read_obj_id", state.get(obj, "id"))
```

Comment:
I'm getting confused about the meaning / purpose of this feature, can you explain?

Reply:
the ID of the object that this satellite has read (used its instrument on), if any. the goal is for each satellite to have read a particular object

Reply:
added comments about the features